pydantic/main.py

```diff
@@ -1112,6 +1112,11 @@ def __init__(__pydantic_self__, root: RootModelRootType) -> None:  # type: ignor
     def model_construct(cls: type[Model], root: RootModelRootType, _fields_set: set[str] | None = None) -> Model:
         return super().model_construct(root=root, _fields_set=_fields_set)
 
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, RootModel):
+            return NotImplemented
+        return self.model_fields['root'].annotation == other.model_fields['root'].annotation and super().__eq__(other)
```

Comment:
My thought was that this should be true `RootModel[int](42) != RootModel[float](42)`, shouldn't it?

Reply:
I think having that return `True` is a price I'm willing to pay to get equality with `Any`. I mean, if you did:

```python
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class MyModel(BaseModel, Generic[T]):
    x: T


assert MyModel[int](x=42) == MyModel[float](x=42.0)
```
it currently doesn't error. And I think there's a lot of stuff that behaves worse if you try to change this, because of things like reparametrizing generics, etc. I just think it's easiest if we have RootModel behave consistently with this

Reply:
@samuelcolvin we also should compare private fields and that is why there is the `super` call.

Reply:
I think comparing private fields is a good reason to keep the `super().__eq__`.

But also, you need to have the `super().__eq__` to make sure the classes match — if you subclass `RootModel[T]` (which I would imagine is the reason why you are using a RootModel and not TypeAdapter in most cases), instance equality should require type equality. At least if we want to be consistent with BaseModel, and I think we should.

I still think we should drop the annotation comparison thing because it's surprisingly inconsistent with other generic BaseModel subclasses, but it's not a hill I want to die on.